docs/design/git-submodules.md

```diff
@@ -137,7 +137,74 @@ I.e. outcomes we would like to see if there were no constraints whatsoever.
+  interact with a submodule. This means that properties that the superproject
+  relies on may be invalidated unexpectedly.
+
+  For example, when `git gc` attempts to delete 'unreachable' objects, it takes
+  into account the objects reachable by refs in the repository. However, in a
+  submodule, it is possible that a commit is not reachable by a submodule ref,
+  but it is reachable by a superproject commit. Without knowledge of the
+  superproject, the submodule may delete objects that the superproject is
+  relying on ([see relevant StackOverflow
+  question](https://stackoverflow.com/questions/31640270/will-git-garbage-collect-commit-in-submodule-referred-to-by-a-top-level-reposito)).
+
+- With Git submodules, 'the repository' changes based on where `git` is run in
+  the working tree, which is a source of
+  [confusion](https://github.com/martinvonz/jj/issues/494#issuecomment-1404338917).
+
+We should keep in mind that a submodule exists to be integrated with a
+superproject, otherwise, the submodule could just be an independent clone. As
+such, jj should enforce tight integrations between superproject and submodule
+tightly by requiring that all interactions with the submodule must be initiated
+from the superproject.
+
+#### Commands should involve submodules by default
+
+Git does not recurse into submodules by default (it requires the
+`--recurse-submodules` flag or the `submodule.recurse` config setting), and this
+is a constant source of confusion for users, e.g.:
+
+- Non-recursive `git clone` will not clone submodules, resulting in their
+  contents being unpopulated in the working tree.
+
+- Non-recursive `git checkout` does not update the submodule index or working
+  tree, resulting in the submodule being 'dirty' and the working tree being out
+  of sync with the superproject.
+
+We should always consider submodules. If they are ignored, we should be explicit
+to the user.
+
+#### Submodules are globally managed
+
+In a regular jj or Git repository, objects are reusable because they are stored
+in a central database, regardless of the working copy. Similarly, we should
+manage submodules globally and treat that as the source of truth; the working
+copy should, at best, be treated as a hint.
+
+A consequence of this is that it is expected that the working copy may be out
+of sync with the global submodules, e.g. submodules may be missing, their
+commits may not be fetched, and they may have different configurations at
+different points in history. This is different from Git, which tends to
+assume that both are always in sync, and behaves badly when they are not. We
+should be prepared to handle missing submodules/submodule commits.
+
+For this to make sense to users, we must make it easy to manage the global
+submodules and to reconcile them with the working copy submodules. Examples of
+this include: prompting the user to update submodules when the working copy
```

Comment:
Why not update submodules automatically when the working copy changes? Actually, which direction do you mean here? Is this about updating the submodule's directory in the working copy or about creating a submodule commit when files have changed in the working copy?

Reply:
> Why not update submodules automatically when the working copy changes?

Ah, I considered this but I couldn't find a good way to weave in "Alternatives considered" in a section so short.

When I say "update submodules", I primarily mean "changing the globally configured set of submodules", not "fetching new commits in already-configured submodules". (Seems like I need to make this clearer.) Automatically updating the current set of submodules to match the working copy makes intuitive sense (and we may want to offer this as optional behavior), but there are a lot of corner cases where it doesn't work well:

- If branches A and B have a disjoint set of submodules, switching between them can delete submodules, causing submodules to need to be cloned over and over. Maybe that's okay if we make this operation append-only, but...
- Git Submodules have global names that identify them across commits (e.g. to track renames). These names can "conflict" in ways that aren't easily resolved.
  - Updating a URL because the submodule's repo changed hosts. Without automatic updates, this is a non-issue because you'd have cloned the submodule from a more recent commit, which should give you the old commits too. With automatic updating, you'd have to know to ignore the old URL somehow.
  -  Two completely different submodules sharing the same "name" at different points in history, e.g. using the name `ui` to refer to the UI framework and then changing UI frameworks at some point. (That's probably a misconfiguration, but it's not nice to punish the user for the mistake forever.) It's not clear to me how automatic updates could solve this, especially if we don't want to throw away submodules. Maybe this will be tractable if we can teach the submodules how to handle conflicts (like we do with branches), but that doesn't seem necessary in the short term.